rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -136,42 +156,268 @@ def consume_result(self, used_amount: FVal) -> None:
+                ),
+            )
+            taxable_amount = adjusted_amount
+            is_complete = False
+
+        return CostBasisInfo(
+            taxable_amount=taxable_amount,
+            taxable_bought_cost=taxable_bought_cost,
+            taxfree_bought_cost=taxfree_bought_cost,
+            matched_acquisitions=matched_acquisitions,
+            is_complete=is_complete,
+        )
 
     def __len__(self) -> int:
         return len(self._acquisitions)
 
 
-class FIFOAcquisitionsOrder(BaseAcquisitionsOrder):
-    """Accounting in FIFO (first-in-first-out) order"""
+class FIFOCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in FIFO (first-in-first-out) method"""
+    _count: int
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = 0
+
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.append(acquisition)
+        heapq.heappush(self._acquisitions, (self._count, acquisition))
+        self._count += 1
+
 
+class LIFOCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in LIFO (last-in-first-out) method"""
+    _count: int
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = 0
 
-class LIFOAcquisitionsOrder(BaseAcquisitionsOrder):
-    """Accounting in LIFO (last-in-first-out) order"""
     def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
-        self._acquisitions.appendleft(acquisition)
+        heapq.heappush(self._acquisitions, (-self._count, acquisition))
+        self._count += 1
+
+
+class HIFOCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in HIFO (highest-in-first-out) method"""
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        heapq.heappush(self._acquisitions, (-acquisition.amount, acquisition))
+
+
+class AverageCostBasisMethod(BaseCostBasisMethod):
+    """Accounting in Average Cost Base(ACB) method"""
+    # the average cost base of last event(buy/sell) that occurred
+    current_average_cost_base: FVal
+    # keeps track of the amount of the asset remaining after evert acquisition or spend
+    remaining_amount: FVal
+    _count: int
+
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = 0
+        self.remaining_amount = ZERO
+        self.current_average_cost_base = ZERO
+
+    def add_acquisition(self, acquisition: AssetAcquisitionEvent) -> None:
+        """
+        This method appends an acquisition to the acquisitions list in FIFO order.
+        It also calculates the average cost base of that acquisition with respect to the
+        previous average cost base.
+
+        The formula used to calculate the average cost base of an acquisition is:
+        [Previous Total ACB] + [Cost of New Shares] + [Transaction Costs]
+        """
+        heapq.heappush(self._acquisitions, (self._count, acquisition))
+        self.current_average_cost_base += (acquisition.rate * acquisition.amount)
```

Comment:
shouldn't the count also be somehow taken into account in the calculation here?

Reply:
Not at all, the count is only used for priority for the heap